messaging/src/main/java/io/lantern/messaging/metadata/Metadata.kt

```diff
@@ -0,0 +1,117 @@
+            return Metadata(mimeType, thumbnail, thumbnail.let { "image/webp" })
+        }
+
+        /**
+         * This is needed to fix orientation of the thumbnails on some phones.
+         * See https://github.com/google/cameraview/issues/22#issuecomment-363047917.
+         */
+        private fun rotatedBitmap(file: File): Bitmap? {
+            val bmp = BitmapFactory.decodeFile(file.absolutePath)
+            if (bmp == null || Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
+                return bmp
+            }
+            val exif = ExifInterface(FileInputStream(file))
+            val orientation: Int = exif.getAttributeInt(
+                ExifInterface.TAG_ORIENTATION,
+                ExifInterface.ORIENTATION_UNDEFINED
+            )
+            val matrix = Matrix()
+            when (orientation) {
+                ExifInterface.ORIENTATION_NORMAL -> return bmp
+                ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> matrix.setScale(-1.0f, 1.0f)
+                ExifInterface.ORIENTATION_ROTATE_180 -> matrix.setRotate(180f)
+                ExifInterface.ORIENTATION_FLIP_VERTICAL -> {
+                    matrix.setRotate(180f)
+                    matrix.postScale(-1f, 1f)
+                }
+                ExifInterface.ORIENTATION_TRANSPOSE -> {
+                    matrix.setRotate(90f)
+                    matrix.postScale(-1f, 1f)
+                }
+                ExifInterface.ORIENTATION_ROTATE_90 -> matrix.setRotate(90f)
+                ExifInterface.ORIENTATION_TRANSVERSE -> {
+                    matrix.setRotate(-90f)
+                    matrix.postScale(-1f, 1f)
+                }
+                ExifInterface.ORIENTATION_ROTATE_270 -> matrix.setRotate(-90f)
+            }
+            val bmpRotated: Bitmap = Bitmap.createBitmap(
+                bmp,
+                0,
+                0,
+                bmp.getWidth(),
+                bmp.getHeight(),
+                matrix,
+                true
+            )
+            return bmpRotated
+        }
+
+        private fun scaledThumbnail(
+            bmp: Bitmap?,
+            maxHeight: Int = 1000,
+            maxWidth: Int = 1000
+        ): ByteArray? {
+            if (bmp == null) {
+                return null
+            }
+            val targetRatio = maxHeight.toFloat() / maxWidth.toFloat()
+            val actualRatio = bmp.height.toFloat() / bmp.width.toFloat()
+            val scaled = if (bmp.height <= maxHeight && bmp.width <= maxWidth)
+                bmp
+            else if (actualRatio > targetRatio)
+            // scale by height
+                bmp.scale((maxHeight / actualRatio).toInt(), maxHeight)
+            else
+                bmp.scale(maxWidth, (maxWidth * actualRatio).toInt())
+            val out = ByteArrayOutputStream()
+            // We use WEBP for backward compatibility with older Android versions
+            scaled.compress(Bitmap.CompressFormat.WEBP, 90, out)
```

Comment:
It also looks like `Bitmap.CompressFormat.WEBP` has been deprecated in API Level 30 in favor of the more explicit `CompressFormat.WEBP_LOSSY` or `CompressFormat.WEBP_LOSSLESS`: https://developer.android.com/reference/android/graphics/Bitmap.CompressFormat

Reply:
Unfortunately those newer formats are only supported on newer devices (API 30 and up), so to use them I'd have to put in a conditional based on the current device version, which seems unnecessary since just using WEBP with quality less than 100 consistently results in lossy compression, which is what I want here.